src/hardware/gameblaster.cpp

```diff
@@ -126,53 +200,39 @@ static uint8_t read_cms_detect(io_port_t port, io_width_t)
 	return retval;
 }
 
-class CMS final : public Module_base {
-private:
-	IO_WriteHandleObject WriteHandler = {};
-	IO_WriteHandleObject DetWriteHandler = {};
-	IO_ReadHandleObject DetReadHandler = {};
-
-public:
-	CMS(Section *configuration) : Module_base(configuration)
-	{
-		Section_prop * section = static_cast<Section_prop *>(configuration);
-		Bitu sampleRate = section->Get_int( "oplrate" );
-		cmsBase = static_cast<io_port_t>(section->Get_hex("sbbase"));
-		WriteHandler.Install(cmsBase, write_cms, io_width_t::byte, 4);
-
-		// A standalone Gameblaster has a magic chip on it which is
-		// sometimes used for detection.
-		const char * sbtype=section->Get_string("sbtype");
-		if (!strcasecmp(sbtype,"gb")) {
-			DetWriteHandler.Install(cmsBase + 4u, write_cms_detect, io_width_t::byte, 12);
-			DetReadHandler.Install(cmsBase, read_cms_detect, io_width_t::byte, 16);
-		}
+void GameBlaster::Close()
+{
+	if (!is_open)
+		return;
 
-		/* Register the Mixer CallBack */
-		cms_chan = MIXER_AddChannel(CMS_CallBack, sampleRate, "CMS");
+	const auto card_name = is_standalone_gameblaster ? "GAMEBLASTER" : "CMS";
+	LOG_INFO("%s: Shutting down the card on port %xh", card_name, base_port);
 
-		lastWriteTicks = PIC_Ticks;
+	// Drop access to the IO ports
+	for (auto &w : write_handlers)
+		w.Uninstall();
+	write_handler_for_detection.Uninstall();
+	read_handler_for_detection.Uninstall();
 
-		machine_config config;
-		device[0] = new saa1099_device(config, "", 0, GAMEBLASTER_CLOCK_HZ);
-		device[1] = new saa1099_device(config, "", 0, GAMEBLASTER_CLOCK_HZ);
+	// Stop playback
+	if (channel)
+		channel->Enable(false);
 
-		device[0]->device_start();
-		device[1]->device_start();
-	}
+	// Remove the mixer channel and the SAA1099 devices
+	channel.reset();
+	devices[0].reset();
+	devices[1].reset();
 
-	~CMS() {
-		cms_chan = 0;
-		delete device[0];
-		delete device[1];
-	}
-};
+	is_enabled = false;
+	is_open = false;
+}
 
-static CMS* test;
-   
-void CMS_Init(Section* sec) {
-	test = new CMS(sec);
+GameBlaster gameblaster;
+void CMS_Init(Section *configuration)
+{
+	Section_prop *section = static_cast<Section_prop *>(configuration);
+	gameblaster.Open(section->Get_hex("sbbase"), section->Get_string("sbtype"));
```

Comment:
Maybe an opportunity to extract some config params referenced in many different places to string constants? (`sbbase` & `sbtype`)

Reply:
DOSBox's init system bring up devices based on conf chunks, so our values come in from the conf system's `[sblaster]` section. 

We could store them in some globals (and use them again in `soundblaster.cpp`) I suppose.  I do like that we're getting them direct from the source - and there's no risk of order-inversion with getting a global value earlier than whatever is response for setting the global value.

Reply:
I agree on that, but I only meant storing the config keys as constants, and only reference them instead of duplicating them many times in different place:

<img width="877" alt="image" src="https://user-images.githubusercontent.com/698770/167538542-8c688dbb-8cd2-41c1-b6e0-6692b21c137f.png">

Reply:
Sorry, I see now. 
Yeah, non-constant strings are currently used throughout the codebase: `git grep 'Get.*("'`.

I like the idea of bakinf them out and using constants (ideally not even strings at that point, could use a struct that has an enum ID and a constexpr string_view to the text key, and perhaps other attributes), I would think it would fall under a separate PR, and we'd adopt it across the board.

Interesting idea! 👍